src/config.rs

```diff
@@ -1,33 +1,140 @@
     // than 2^16 proc at any single time.
     numprocs: u16,
+    /// Program mask running under
+    umask: i32,
     /// Working directory of the task.
     workingdir: PathBuf,
+    /// Service starting at load time.
+    autostart: bool,
+    /// Service restarting when finished.
+    autorestart: Restart,
+    /// Autorestart callback trigger value.
+    exitcodes: Vec<i32>,
+    /// Restart attempts.
+    startretries: i32,
+    /// Delay before start.
+    starttime: i32,
+    /// Signal received by supervisord trigger service termination.
+    stopsignal: String,
+    /// Running period.
+    stoptime: i32,
+    /// Stdout logfile path.
+    stdout: PathBuf,
+    /// Stderr logfile path.
+    stderr: PathBuf,
+    /// Environment variable service is running into.
+    env: HashMap<String, String>,
 }
 
 impl Default for Config {
     fn default() -> Self {
+        let bin = "ls";
+        let home = dirs::home_dir();
+        if home.is_none() {
+            eprintln!("Impossible to get user home directory!");
+            process::exit(1);
+        }
+        let home = home.unwrap();
+
+        #[cfg(debug_assertions)]
+        println!("{:?}", home);
+
         Self {
-            cmd: "ls".into(),
+            cmd: bin.into(),
             numprocs: 1,
-            workingdir: PathBuf::from("/tmp"),
+            umask: 0o027,
+            workingdir: home.clone(),
+            autostart: true,
+            autorestart: Restart::Never,
+            exitcodes: {
+                let mut v = Vec::new();
+                v.push(0);
+                v
+            },
+            startretries: 0,
+            starttime: 0,
+            stopsignal: "SIGTERM".to_string(),
+            stoptime: 0,
+            stdout: home.join(bin.to_string() + ".stdout"), // Would be lovely to replace "bin" by "cmd" from the struct itself
+            stderr: home.join(bin.to_string() + ".stderr"), // Same
+            env: HashMap::new(),
         }
     }
 }
 
-impl std::convert::TryFrom<&Path> for Config {
-    type Error = std::io::Error;
+/// # Errors
+///
+/// Will return `Err` if `file` does not exist or the user does not have
+/// permission to read it or yaml format is not respected.
+pub fn parse(
+    file_name: &Path,
+) -> Result<HashMap<String, HashMap<String, Config>>, serde_yaml::Error> {
+    let file = File::open(&file_name).expect("Unable to open config file");
+    let d: HashMap<String, HashMap<String, Config>> = serde_yaml::from_reader(file)?;
+    Ok(d)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn empty_default_values() {
+        let conf = Config::default();
+        assert_eq!(conf.cmd, "ls");
+        assert_eq!(conf.exitcodes[0], 0);
+        assert_eq!(conf.workingdir, dirs::home_dir().unwrap());
+        assert_eq!(conf.stdout, dirs::home_dir().unwrap().join("ls.stdout"));
+        assert_eq!(conf.stderr, dirs::home_dir().unwrap().join("ls.stderr"));
```

Comment:
Here we're just testing that the `default` function works. I'm not sure it's really relevant ; Rust is pretty solid and I don't think we should expand code to fail.
However, I think testing the `Serialization` / `Deserialization`, and actually verifying that the fields are correct, could be cool :)

Reply:
Maybe doing something like a `round_trip` could be cool (You take a `Config`, serialize it, deserialize it, and assert that the `Config` you get it identical to the original one) :D